src/DataFactory.js

```diff
@@ -1,182 +1,45 @@
-const FateInt = require('./types/FateInt')
-const FateBool = require('./types/FateBool')
-const FateString = require('./types/FateString')
-const FateHash = require('./types/FateHash')
-const FateSignature = require('./types/FateSignature')
-const FateList = require('./types/FateList')
-const FateMap = require('./types/FateMap')
-const FateTuple = require('./types/FateTuple')
-const FateVariant = require('./types/FateVariant')
-const FateBytes = require('./types/FateBytes')
-const FateBits = require('./types/FateBits')
-const FateAccountAddress = require('./types/FateAccountAddress')
-const FateContractAddress = require('./types/FateContractAddress')
-const FateChannelAddress = require('./types/FateChannelAddress')
-const FateOracleAddress = require('./types/FateOracleAddress')
-const FateOracleQueryAddress = require('./types/FateOracleQueryAddress')
+const PrimitiveDataFactory = require('./DataFactory/PrimitiveDataFactory')
+const ListDataFactory = require('./DataFactory/ListDataFactory')
+const MapDataFactory = require('./DataFactory/MapDataFactory')
+const TupleDataFactory = require('./DataFactory/TupleDataFactory')
+const RecordDataFactory = require('./DataFactory/RecordDataFactory')
+const VariantDataFactory = require('./DataFactory/VariantDataFactory')
 
 const zip = (arr, ...arrs) => {
     return arr.map((val, i) => arrs.reduce((a, curr) => [...a, curr[i]], [val]))
 }
 
 class DataFactory {
-    constructor(aci) {
-        this.aci = aci
+    #factories
+
+    #mapper
+
+    constructor(mapper) {
+        this.#mapper = mapper
+        this.#factories = [
+            new PrimitiveDataFactory(this),
+            new ListDataFactory(this),
+            new MapDataFactory(this),
+            new TupleDataFactory(this),
+            new RecordDataFactory(this),
+            new VariantDataFactory(this),
+        ]
     }
 
     create(types, values) {
-        const requiredTypesCount = types
-            .map(type => FateVariant.isFateTypeOption(type))
-            .lastIndexOf(false) + 1
-        if (values.length > types.length || values.length < requiredTypesCount) {
-            throw new Error(
-                'Non matching number of arguments. '
-                + `Got ${values.length} but expected from ${requiredTypesCount} to ${types.length}`
-            )
```

Comment:
Currently, I have error:
> Variant should be an object mapping constructor to array of values, got "undefined" instead

instead of
> Non matching number of arguments. Got 0 but expected from 1 to 1"

This check was clearly explaining the case of wrong arguments count, I'm for keeping it

Reply:
I just forgot to re-add it